operators/pkg/instance-controller/controller_test.go

```diff
@@ -12,290 +12,408 @@
-						Memory:                resource.MustParse("1024M"),
-					},
-					EnvironmentType: crownlabsv1alpha2.ClassVM,
-					Persistent:      false,
-					Image:           "trololo/vm",
-				},
-			},
-			DeleteAfter: "",
+		if createWorkspace {
+			Expect(k8sClient.Create(ctx, &workspace)).ToNot(HaveOccurred())
 		}
-		tenant = crownlabsv1alpha1.Tenant{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: TenantName,
-			},
-			Spec: crownlabsv1alpha1.TenantSpec{
-				FirstName:  "Mario",
-				LastName:   "Rossi",
-				Email:      "mario@rossi.com",
-				Workspaces: nil,
-				PublicKeys: []string{
-					"ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDP5O4CX17GK17GN+xxfoUpjz6s1FLLQdVSBYSJ02uS/HHueJmE8TQS8tpNfVHQ+i9cCpR+RMXUUDscNjgTxcF8Z0iRIfX6InRUbJt7FSYERX3roSy4YyhNnkDhQe9+1cQNhZsUtVKkTAE4Ew/dnqHjAKMjz+nhiNio7bL0ZsODGZ90uFUfzcg2RUyftluN+IaX9cD9VvmXRmKFMiIUkebOnLxREiOS6aqe8NrqbkK6Bkt0hWhr8U2pfDK86BbgFGL9e7Ms1dWxDPfrOVLnteN0Xe0bLv3HoW1KVlnoC4hwSHaRlhlSU0wgTkfvPzQy/eM95oTrQQCr0fmvjv5uiciP",
-				},
-				CreateSandbox: false,
-			},
-			Status: crownlabsv1alpha1.TenantStatus{},
+		if createTemplate {
+			Expect(k8sClient.Create(ctx, &template)).ToNot(HaveOccurred())
 		}
-		ns1  = v1.Namespace{}
-		tmp1 = crownlabsv1alpha2.Template{}
-	)
+		Expect(k8sClient.Create(ctx, &instance)).ToNot(HaveOccurred())
+	})
 
-	Context("", func() {
-		It("Setting up the Instance and Template namespaces", func() {
-			ctx = context.Background()
-			Expect(k8sClient.Create(ctx, &templateNs)).Should(Succeed())
-			doesEventuallyExist(ctx, types.NamespacedName{
-				Name: TemplateNamespace,
-			}, &ns1, BeTrue(), timeout, interval)
-			Expect(k8sClient.Create(ctx, &instanceNs)).Should(Succeed())
-			Expect(k8sClient.Create(ctx, &webdavSecret)).Should(Succeed())
-			doesEventuallyExist(ctx, types.NamespacedName{
-				Name: InstanceNamespace,
-			}, &ns1, BeTrue(), timeout, interval)
-			Expect(k8sClient.Create(ctx, &tenant)).Should(Succeed())
-			doesEventuallyExist(ctx, types.NamespacedName{
-				Name: TenantName,
-			}, &tenant, BeTrue(), timeout, interval)
-		})
-		Context("Create an Instance from a single VM template", func() {
-			It("Creates the single-vm Template and the instance", func() {
-				By("Creating the Template")
-				template := crownlabsv1alpha2.Template{
-					TypeMeta: metav1.TypeMeta{},
-					ObjectMeta: metav1.ObjectMeta{
-						Name:      TemplateName,
-						Namespace: TemplateNamespace,
-					},
-					Spec:   templateSpecSingleVM,
-					Status: crownlabsv1alpha2.TemplateStatus{},
+	Context("The instance is container based", func() {
+		When("the environment is persistent", func() {
+			BeforeEach(func() {
+				testName = "test-container-persistent"
+				environment = clv1alpha2.Environment{
+					Name:            testName,
+					Image:           "docker-image:v0",
+					EnvironmentType: clv1alpha2.ClassContainer,
+					GuiEnabled:      true,
+					Resources:       testResources,
+					Mode:            clv1alpha2.ModeStandard,
+					Persistent:      true,
 				}
```

Comment:
Nit. Could move a level up to reduce duplication, and just overwrite the necessary fields.